### ISSUE TITLE
StackOverflow when implementing a resource locator which will return the resource class itself

### ISSUE THREAD

**Issue by fabiobrz** [smallrye/smallrye-open-api#353]
Hi, I've noticed that some RESTEasy tests were failing against Wildfly because of some resource classes which have resource locators that are supposed to return the resource class itself - i.e. `this`.
Of course the annotation scanner goes crazy until the stack overflow happens because it tries to process the JAX-RS resource class recursively but I was wondering whether you think that this is an issue and that the annotation scanner should be protected against this case. WDYT?

The following stack trace represents the output of a deployment failure when an attempt to deploy a WAR with such a resource:

```
at org.jboss.logging@3.4.1.Final-redhat-00001//org.jboss.logging.Logger.getLogger(Logger.java:2490)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.dataobject.DataObjectDeque.<init>(DataObjectDeque.java:38)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.<init>(OpenApiDataObjectScanner.java:138)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:163)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.util.SchemaFactory.typeToSchema(SchemaFactory.java:360)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.createResponseFromJaxRsMethod(OpenApiAnnotationScanner.java:924)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsMethod(OpenApiAnnotationScanner.java:688)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.lambda$processJaxRsResourceClass$0(OpenApiAnnotationScanner.java:422)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1603)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsResourceClass(OpenApiAnnotationScanner.java:420)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsSubResource(OpenApiAnnotationScanner.java:511)
        at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsResourceClass(OpenApiAnnotationScanner.java:426)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsSubResource(OpenApiAnnotationScanner.java:511)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsResourceClass(OpenApiAnnotationScanner.java:426)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsSubResource(OpenApiAnnotationScanner.java:511)

...
```

I've isolated a reproducer [1] which has one method (`FruitResource::getResourceLocator()`) to demonstrate the failure.
The reproducer should fail by default. For a successful deploy the above mentioned method method should be commented.

The SmallRye OpenAPI versions that I tested this against are 1.1.21 and 1.1.22.
 
[1]
https://github.com/fabiobrz/microprofile-openapi-example/tree/subresource-locator-repro

**Comment by MikeEdgar** [comment]
@fabiobrz - I agree, it makes sense for the scanner to guard against this scenario.

**Comment by fabiobrz** [comment]
Hi @MikeEdgar, do you think this fix could be backported to 1.1.x as well?

**Comment by MikeEdgar** [comment]
@fabiobrz, yes certainly. Are there any other issues in 1.1.x that you are working on? I'd like to bundle them to reduce the overhead of making changes on multiple branches, etc.

**Comment by fabiobrz** [comment]
Hi @MikeEdgar, yeah I perfectly get your point.
I've been taking care of RESTEasy TS failures which popped up since we started executing those tests against a Wildfly server instance with enabled MP OpenAPI subsystem. 
Looking at the list of the failing items I don't have anything else at the moment.

**Comment by MikeEdgar** [comment]
Thanks @fabiobrz. I haven't had time to do the research, but do you know how RESTEasy handles this type of configuration? In [your example](https://github.com/fabiobrz/microprofile-openapi-example/blob/subresource-locator-repro/src/main/java/com/example/microprofile/openapi/FruitResource.java), would all of the endpoints under `/fruit` also be available under `/fruit/resource`?

**Comment by fabiobrz** [comment]
HI @MikeEdgar,
> I haven't had time to do the research, but do you know how RESTEasy handles this type of configuration?
RESTEasy has got a couple of tests failing for this, namely https://github.com/resteasy/Resteasy/blob/3.11.2.Final/testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/core/smoke/WireSmokeTest.java which is using [1] and https://github.com/resteasy/Resteasy/blob/3.11.2.Final/testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/path/LocatorSubResourceReturningThisTest.java which in turn is using its internal [2].

That said, I have been going through some doc here: https://github.com/resteasy/Resteasy/blob/3.11.2.Final/docbook/reference/en/en-US/modules/JAX-RS_Resource_Locators_and_Sub_Resources.xml
which leads me to the assumption that this kind of configuration would result in a recursive routing under `resource`, in my repro. 
That is the reason why I actually think that the current behavior of the annotation scanner seems to be legitimate, although in need for a guard against the stack overflow race condition as you stated.
Does this make sense? 


[1]
https://github.com/resteasy/Resteasy/blob/3.11.2.Final/testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/core/smoke/resource/WireSmokeLocatingResource.java

[2]
https://github.com/resteasy/Resteasy/blob/3.11.2.Final/testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/path/LocatorSubResourceReturningThisTest.java#L38

**Comment by fabiobrz** [comment]
> Let's hope I am not missing anything :)

BTW @MikeEdgar I see there's another test which is failing due to a NPE and it is https://github.com/resteasy/Resteasy/blob/3.11.2.Final/testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/form/Resteasy1405Test.java.

I'll have to see what's the reason for that, though, because I am failing at binding it to the root causes that I have reported previously, at the moment.

**Comment by MikeEdgar** [comment]
@fabiobrz - I agree that the behavior is generally legitimate. I was thinking about allowing the scanner to recurse one level before stopping. If RestEasy allows recursing without a guard as well, it seems like a runtime vulnerability.

**Comment by fabiobrz** [comment]
> I agree that the behavior is generally legitimate. I was thinking about allowing the scanner to recurse one level before stopping.
If RestEasy allows recursing without a guard as well, it seems like a runtime vulnerability.

Sure, I get your point @MikeEdgar and the 1-level iteration would seem the most natural implication of such a configuration, to me.
That said, yeah, I am on the same page on that last consideration of yours but couldn't actually do the research which could help me get how RESTEasy would behave. 
Maybe @ronsigal can help and share some knowledge over this particular scenario?

**Comment by MikeEdgar** [comment]
@fabiobrz - did you get a chance to look at the NPE in the other test?

**Comment by MikeEdgar** [comment]
@fabiobrz - if you determine that NPE to be a separate problem, go ahead and open up a new issue. I went ahead and resolved the stack overflow issue for now.

**Comment by fabiobrz** [comment]
Hi @MikeEdgar. Thanks a lot for your feedback. 
I am still looking after it. It keeps on failing in our CI runs but it passes locally. I failed to create a reproducer, as the one I made using RESTEasy resources has been deployed successfully. So, still looking for the root cause.
I investigated the stack trace as well and got to this bit:
```
Caused by: java.lang.NullPointerException
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.ParameterProcessor.setSchemaProperties(ParameterProcessor.java:637)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.ParameterProcessor.getFormBodyContent(ParameterProcessor.java:605)
```
Here's where it fails, then: https://github.com/smallrye/smallrye-open-api/blob/1.1.22/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/ParameterProcessor.java#L605

**Comment by MikeEdgar** [comment]
@fabiobrz - looks like the same stack trace as #349 . Are you running a SmallRye SNAPSHOT locally that would already contain the fix?

**Comment by fabiobrz** [comment]
Hi @MikeEdgar, you're right, it _seems_ but it is not. 
The fact is that when filing that issue (#349) I copied the stack trace just below that very test or the other one that fails with the same stack trace.
I hadn't realized it was different for all the other tests involving `SortedSet`, yet.

The ones failing because of `SortedSet` have the following stack trace:
```
Caused by: java.lang.NullPointerException
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.ParameterProcessor.setSchemaProperties(ParameterProcessor.java:637)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.ParameterProcessor.getParameters(ParameterProcessor.java:506)
```

**Comment by fabiobrz** [comment]
> ... Are you running a SmallRye SNAPSHOT locally that would already contain the fix?
About to try this, thanks. It might be fixed already :)

**Comment by fabiobrz** [comment]
Hi @MikeEdgar, so finally here I am :)
I've been trying 1.1.23-SNAPSHOT and it actually confirmed how the issue with this specific test differs from the others, which seem to be fixed.
Some context: I am referring to the following test: https://github.com/fabiobrz/Resteasy/blob/3.11-removing-annotation-to-skip-tests-openapi-issues/testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/form/Resteasy1405Test.java
As said I failed in reproducing it the same way I could with previous tests - i.e. build a WAR based on same logic/resource types. 
This means that if I build a WAR out of the same resources and data types that the test is using and deploy it to the server, well... no failure at all. So I asked why the Arquillian deployment in RESTEasy TS is failing instead.
I eventually ended up in having both the WARs (the one built by myself with the est resources and the one generated by Arquillian during RESTEasy test suite execution) and step-by-step debugging the deployment operation.
Bottom line - when deplying the WAR generated by Arquillian it fails here:
https://github.com/smallrye/smallrye-open-api/blob/1.1.x/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/ParameterProcessor.java#L641
being `paramSchema` - computed at https://github.com/smallrye/smallrye-open-api/blob/1.1.x/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/ParameterProcessor.java#L638 - just _null_, hence the NPE.
I still don't get why that variable is _null_ for the Arquillian deployment case and instead `InputPart` (as expected) for the manually generated WAR but it has to do with the `index` passed here, which stores completely different data and eventually leads https://github.com/smallrye/smallrye-open-api/blob/1.1.x/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/OpenApiDataObjectScanner.java#L201 to return `null`.
The fact that a similar (manually assembled) WAR could be deployed successfully would make me think that this is not a SmallRye OpenAPI issue, bbut I'd like to hear your opinion on this.
Besides, maybe it could be a good idea to protect api/blob/1.1.x/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/ParameterProcessor.java#L641 from the above described condition (i.e.: assert paramSchema != null).
WDYT?
